enclave-runner/src/usercalls/mod.rs

```diff
@@ -346,36 +381,204 @@ impl EnclaveState {
             fds: Mutex::new(fds),
             last_fd,
             exiting: AtomicBool::new(false),
-            usercall_ext : usercall_ext,
+            usercall_ext,
         })
     }
 
+    fn create_worker_threads(num_of_worker_threads: usize,
+                             work_receiver: crossbeam::crossbeam_channel::Receiver<Work>,
+                             io_queue_send: mpsc::Sender<UsercallSendData>
+    ) -> Vec<JoinHandle<()>> {
+        let mut thread_handles = vec![];
+        for _ in 0..num_of_worker_threads {
+            let work_receiver = work_receiver.clone();
+            let io_queue_send = io_queue_send.clone();
+
+            thread_handles.push(thread::spawn(move || {
+                while let Ok(work) = work_receiver.recv() {
+                    work.do_work(&io_queue_send);
+                }
+            }));
+        }
+        return thread_handles;
+    }
+
+    fn  syscall_loop(io_queue_receive: mpsc::Receiver<UsercallSendData>, enclave: Arc<EnclaveState>,
+                     work_sender: crossbeam::crossbeam_channel::Sender<Work>
+    ) -> StdResult<(u64, u64), EnclaveAbort<EnclavePanic>> {
+        while let Some((coresult, mut state, buf)) = io_queue_receive.recv().ok() {
+            let mut erased_tcs = None;
+            let enclave = enclave.clone();
+            let my_result = match coresult {
+                CoResult::Yield(usercall) => {
+                    let result;
+                    {
+                        let mut input = IOHandlerInput { enclave: &enclave, tcs: &mut state, work_sender: &work_sender};
+                        let (p1, p2, p3, p4, p5) = usercall.parameters();
+                        result = dispatch(&mut Handler(&mut input), p1, p2, p3, p4, p5);
+                    }
+                    match result {
+                        Ok(ret) => {
+                            work_sender
+                                .send(Work {tcs:state, entry: CoEntry::Resume(usercall, ret )})
+                                .expect("Work sender couldn't send data to receiver");
+                            continue;
+                        },
+                        Err(EnclaveAbort::Exit { panic: true }) => {
+                            #[cfg(unix)]
+                            trap_attached_debugger(usercall.tcs.address().0 as _);
```

Comment:
It's ok to add a method to `Usercall` to get the TCS address.

Reply:
I'm trying to do that.
```    
impl<T: Tcs> Usercall<T> {
    ...
    pub fn tcs_address(&self) -> *mut c_void { self.tcs.address() }
}
```
However for some reason, I'm having to change
`trap_attached_debugger(usercall.tcs.address().0 as _);`
to 
`trap_attached_debugger(usercall.tcs_address() as _);`

This is confusing me.